Add enet functionality

johnlees opened the pull request:
This adds the ability to fit and predict with the elastic net into pyseer. Some notes:
- Currently should be considered experimental, but mature enough to merge into master for the next release.
- I have tested against equivalent R code, getting the same results.
- I have updated the docs accordingly.
- I have not yet checked or updated the tests, I will do this as soon as possible.
- The `glmnet` python port is most reliably installed by a conda recipe I wrote, rather than through `pip` (which only works out of the box using an intel-based linux platform, due to a packaged pre-compiled object file). I've updated `setup.py` to work out of the box where possible, but I have changed the documentation to prefer the conda route from now on. Once merged and released, I'll make a PR to update the conda recipe too.

@mgalardini if you have free time could you quickly browse through and check I haven't made any obvious clangers or terrible stylistic choices? No worries if you are busy!
I will go through the tests in detail myself before merging

johnlees commented:
That is more commits than I thought!

mgalardini commented:
Hi John!

This is great! I'm travelling until the 28th unfortunately, but if this can
wait a bit I will be happy to review/add tests/update the docs.

M

On Thu, 18 Oct 2018, 17:02 John Lees, <notifications@github.com> wrote:

> That is more commits than I thought!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mgalardini/pyseer/pull/58#issuecomment-431179382>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABErX9Ou1MGa4WIYpne_-t8npCiXssVNks5umPqMgaJpZM4XvUap>
> .
>

johnlees commented:
It certainly can wait! I'll try and minimise the amount that needs to be added/fixed before then too.
Enjoy your trip!

johnlees commented:
I've moved the `travis.yaml` setup to use conda for the `glmnet_py` dependency and fixed the existing tests. The only concern that came up was a change in the LMM using covariates. I suspect we might have been doing it wrong before and this fixes it (also one user told me they got no differences including the covariates in the LMM, but did in the fixed effects model), but would appreciate the check.

I have yet to add any tests on the new functionality. We could discuss suitable test data when you have a chance.

johnlees commented:
> That is more commits than I thought!

We can squash when we merge

mgalardini commented:
Hi John,

I'm now back online and ready to have a look into this. I'll start poking around to test the new functionality, than I'll review the code and tests. I hope to be done this week. Please let me know if I can do anything more specific!

johnlees commented:
Ok, that sounds great! I've probably got one more commit to fix a problem, which I'll make later today
I can send some test data and example commands if that would help

mgalardini commented:
Hi John, yes please if you could send some test data/commands that would be great! I'll do some independent tests on my own data to see if anything else pops up.

johnlees commented:
For testing the elastic net fitting, saving, loading and prediction:
```
python pyseer-runner.py --vcf ma_snps.maf_filtered.vcf.gz --phenotypes ma.pen.pheno --save-enet ma_snps --enet --save-model outputs/ma.pen.lasso --alpha 1 > outputs/ma_pen.lasso.txt
python pyseer-runner.py --vcf ma_snps.maf_filtered.vcf.gz --phenotypes ma.pen.pheno --load-enet ma_snps --enet --save-model outputs/ma.enet.lasso > outputs/ma_enet.lasso.txt
python enet_predict-runner.py --vcf ma_snps.maf_filtered.vcf.gz outputs/ma.enet.lasso.pkl  ma_samples.list > outputs/ma.pen.lasso.predict.txt
```

For testing prediction from the existing models:
```
python pyseer-runner.py --vcf ma_snps.maf_filtered.vcf.gz --phenotypes ma.pen.pheno --lmm --similarity mass_phylogeny_K.tsv > outputs/ma_pen.lmm.txt
python scripts/save_model.py --p-cutoff 1.07E-06 outputs/ma_pen.lmm.txt outputs/ma_pen.lmm
python enet_predict-runner.py --vcf ma_snps.maf_filtered.vcf.gz outputs/ma_pen.lmm_model.pkl ma_samples.list > outputs/ma.pen.lmm.predict.txt
```
Files here:
[enet_test_files.tar.gz](https://github.com/mgalardini/pyseer/files/2534863/enet_test_files.tar.gz)

mgalardini commented:
Hi,

just started looking into this, I'm documenting the various problems I'm finding:

```$ python pyseer-runner.py --vcf test.vcf --phenotypes ma.pen.pheno --save-enet ma_snps --enet --save-model ma.pen.lasso --alpha 1.0
Read 603 phenotypes
Detected binary phenotype
Reading all variants
4986variants [00:05, 845.13variants/s]
Applying correlation filtering
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4986/4986 [00:03<00:00, 1499.82variants/s]
Fitting elastic net to top 3739 variants
Traceback (most recent call last):
  File "pyseer-runner.py", line 8, in <module>
    main()
  File "[...]/pyseer/__main__.py", line 592, in main
    enet_betas = fit_enet(p, all_vars, cov, options.continuous, options.alpha, options.n_folds, options.cpu)
  File "[...]/pyseer/enet.py", line 156, in fit_enet
    nfolds = n_folds, alpha = alpha, parallel = n_cpus)
  File "[...]/venv/lib/python3.6/site-packages/glmnet_python/cvglmnet.py", line 260, in cvglmnet
    ma = scipy.tile(scipy.arange(nfolds), [1, scipy.floor(nobs/nfolds)])
  File "[...]/venv/lib/python3.6/site-packages/numpy/lib/shape_base.py", line 1157, in tile
    return c.reshape(shape_out)
TypeError: 'numpy.float64' object cannot be interpreted as an integer```

I think the problem can be solved by changing this line:

`ma = scipy.tile(scipy.arange(nfolds), [1, scipy.floor(nobs/nfolds)])`

into this other line:

`ma = scipy.tile(scipy.arange(nfolds), [1, scipy.floor(nobs/nfolds).astype(int)])`

But it turns out that the version that is available through `pip` is not the most up to date (https://github.com/bbalasub1/glmnet_python/issues/20). I'll change the requirements file to point to master.

mgalardini commented on pyseer/input.py:
```diff
@@ -359,7 +359,7 @@ def read_variant(infile, p, var_type, burden, burden_regions,
                                                 int(region.group(2)) - 1,
                                                 int(region.group(3))):
-                        var_sub_name = read_vcf_var(variant, d)
+                        var_sub_name = read_vcf_var(variant, d, keep_list)
```
I haven't tried if this change actually works yet, but in theory we should pass the keep list here as well, right?

johnlees replied:
Yes, this makes sense!

mgalardini commented:
Hi John, the `--threshold` option in `enet_predict` refers to the `probability` column if I understand correctly, right? Before the change it was referring to the `link` column.

mgalardini commented on pyseer/enet_predict.py:
```diff
@@ -0,0 +1,183 @@
+    # Note those variants which did not appear - impute
+    for missing in model_dict.keys():
+        sys.stderr.write("Could not find covariate/variant " + missing + " in input file\n")
+        predictions += model_dict[missing][0] * model_dict[missing][1]
```
So if I understood correctly here we are using the AF and the betas to impute missing variants. Why would a variant be missing in the first place? Because the sample file is smaller than the total list of strains? Also, in `__main__.py` the first element might be either AF (line 626) or the average of the covariates (line 632)? Have I got it right?

johnlees replied:
The most likely reason for a variant being missing would be trying to apply a fitted model to perform prediction on a new dataset.
The first element should be the mean of the predictor x i.e. E(x) / <x>. So I think both the AF or average of the covariate would be right

johnlees commented:
> Hi John, the `--threshold` option in `enet_predict` refers to the `probability` column if I understand correctly, right? Before the change it was referring to the `link` column.

Yes, it should be `probability` rather than `link`. I did have a change where I'd mixed these comments up so this change would make sense

mgalardini commented:
Hi John, I think it can be merged, yes! I will try to add some tests too if I can, really sorry for being so slow!

johnlees commented:
Thanks for reviewing! 
Overall it still needs to be added to anyway so maybe we should wait with test, but I think it's a good idea to merge at this point
